src/Castle.Core/DynamicProxy/ModuleScope.cs

```diff
@@ -174,11 +189,94 @@ public Type GetFromCache(CacheKey key)
 		/// </summary>
 		/// <param name = "key">The key to be associated with the type.</param>
 		/// <param name = "type">The type to be stored in the cache.</param>
+		[Obsolete]
+		[EditorBrowsable(EditorBrowsableState.Never)]
 		public void RegisterInCache(CacheKey key, Type type)
 		{
 			typeCache[key] = type;
 		}
 
+		/// <summary>
+		///   Returns a type from this scope's type cache, or registers it in the cache if the key cannot be found.
+		/// </summary>
+		/// <param name="key">The key to be looked up in the cache.</param>
+		/// <param name="valueFactory">A function producing the type to be registered in the cache.</param>
+		/// <returns>The type from this scope's type cache matching the key, or null if the key cannot be found</returns>
+		/// <remarks>
+		///   This method does not synchronizes accesses to this scope's type cache using read/write locks.
+		///   Only use this method when you know for sure that a method further up in the call stack already holds
+		///   a write lock. The function must not access this scope's type cache.
+		/// </remarks>
+		public Type GetFromOrRegisterInCache(CacheKey key, Func<CacheKey, Type> valueFactory)
```

Comment:
Since methods like `RegisterInCache` are being deprecated, are you making this public for a good reason?

I don't see why we'd want to allow users to add their own types to the cache. We've already got `LoadAssemblyIntoCache` for loading a DP generated assembly which is a special case.

Reply:
> I don't see why we'd want to allow users to add their own types to the cache.

Neither do I.

> Since methods like `RegisterInCache` are being deprecated, are you making this public for a good reason?

If we agree that the proxy type cache shouldn't be exposed, I'd be more than happy to make this `internal`. I just didn't know the full story behind *why* the cache is exposed, so I didn't want to just reverse it without further discussion.

Reply:
> If we agree that the proxy type cache shouldn't be exposed, I'd be more than happy to make this internal. I just didn't know the full story behind why the cache is exposed, so I didn't want to just reverse it without further discussion.

The old methods will be gone in the next major version, so we'll hear from someone if they think they have a valid use case (doubt it), but let's not make new stuff public unless we actually want to.